mono/utils/mono-threads.c

```diff
@@ -860,16 +873,32 @@ mono_thread_info_resume (MonoNativeThreadId tid)
 	return result;
 }
 
+static gboolean
+begin_suspend_request_hybrid_suspension (MonoThreadInfo *info)
+{
+	switch (mono_threads_transition_request_hybrid_suspension (info)) {
+	case HybridSuspendAlreadySuspended:
+		return TRUE;
+	case HybridSuspendInitCoopSuspend:
+		return begin_safepoint_suspend (info);
+	case HybridSuspendInitBlockingSuspend:
+		/* signal the thread, but don't abort syscalls */
+		return begin_preemptive_suspend (info, FALSE);
+	default:
+		g_assert_not_reached ();
+	}
+}
+
 gboolean
 mono_thread_info_begin_suspend (MonoThreadInfo *info)
 {
+	if (G_UNLIKELY (mono_threads_is_hybrid_suspension_enabled ())) {
+		return begin_suspend_request_hybrid_suspension (info);
```

Comment:
Let me see if I'm getting this: have a single "begin suspending" transition that subsumes both `request_async_suspension` and `request_hybrid_suspension`;  give it enough return codes so that we can switch on the results and in all cases do the correct `begin_safepoint_suspend`/`begin_preemptive_suspend`?

That makes sense but there are a couple problems:
1. (minor) right now `ASYNC_SUSPENDED` has a `request_async_suspension` selfloop, but no legal `request_hybrid_suspension` transition (if we decided to suspend that thread via safepointing, it coudn't have preemptively suspended and shouldn't try to preemptively suspend for suspend_count >1.  This one is probably harmless to add.).  The symmetric situation on `BLOCKING_ASYNC_SUSPENDED` - it has a `request_hybrid_suspension` selfloop but no legal `request_async_suspension` transition - if we decided to suspend it preemptively we shouldn't try to suspend it again via checkpointing. I think it's better when more things are illegal.
2. (serious) in `BLOCKING` state, `request_async_suspension` loops back to `BLOCKING`, but `request_hybrid_suspension` goes to `BLOCKING_SUSPEND_REQUESTED` - this decision is driven purely by external considerations, not by the state machine state.  So we need two distinct transitions for this case.

In general, the suspend transitions represent *policy* - different suspend techniques starting from the same states want different decisions.  On the resume side it's different - the resume policy is encoded in the state that the thread entered when we suspended it.

Alternately, maybe we could split the blocking state into, let's say `BLOCKING_NONPREEMPTABLE` (kind of like the original meaning of BLOCKING - it's in a syscall or foreign code that doesn't touch managed at all which will suspend at safepoints on the done/abort transitions) and `BLOCKING_PREEMPTABLE` which is preempted.  In that case we push all the complexity into runtime code (we now need to do BEGIN_SYSCALL/END_SYSCALL and BEGIN_FOREGIN/END_FOREIGN).  On the bright side we won't have to restart syscalls.

Reply:
In the morning things look better.  The solution is to do a two-phase suspend in blocking always, the same way that we do it in the running case (ie: `request_suspension` should move from `Blocking` to `Blocking_Suspend_Requested`, always, which communicates enough information to the suspend initiator to then decide on a suspend policy).  The previous `request_async_suspension` transition from `Blocking` to `Blocking` was just an artifact of not having an explicit `Blocking_Suspend_Requested` state.

And there's a benefit for the state machine, too: we can establish the invariant that in `Blocking` the suspend count is always equal to 0 and in `Blocking_Suspend_Requested` it's always strictly positive.

I'm going to update the PR to make it work this way.

Thanks @luhenry !